components/make-static-previews.py

```diff
@@ -0,0 +1,145 @@
+    appfile = rootdir / appfile
+    destfile = rootdir / destfile
+
+    app_mod = importlib.import_module(
+        str(appfile.relative_to(cwd).with_suffix("")).replace("/", ".")
+    )
+    if "app_ui" not in dir(app_mod):
+        raise RuntimeError(f"app_ui not found in {appfile}")
+    app_ui = app_mod.app_ui
+
+    enrich_app_ui(app_ui)
+
+    destfile.parent.mkdir(parents=True, exist_ok=True)
+
+    app_ui.save_html(destfile, libdir=relative_to(libdir, destfile.parent))
+
+
+def enrich_app_ui(app_ui: Tag):
+    """
+    Need to add Shiny dependency because some inputs/outputs won't look right without
+    being intialized as part of the Shiny binding process.
+    """
+    app_ui.append(shiny.html_dependencies.shiny_deps())
+    app_ui.append(
+        head_content(
+            tags.script(
+                """
+                window.Shiny = window.Shiny || {};
+                window.Shiny.createSocket = function(url) {
+                    // Prevent Shiny from trying to connect to a live server
+                    // by returning a dummy WebSocket object that just no-ops
+                    return {
+                        addEventListener: (event, callback) => {},
+                        send: () => {},
+                        close: () => {},
+                        readyState: 0,
+                        url
+                    };
+                };
+                """
+            )
+        )
+    )
+
+
+def relative_to(path1: Path, path2: Path) -> Path:
+    """
+    Equivalent to `path1.relative_to(path2, walk_up=True)`, except that walk_up=True is
+    only supported from Python 3.12 on and this function works on older versions. If and
+    when we can assume Python 3.12, this function can be removed in favor of the
+    built-in method.
+
+    Example:
+    >>> relative_to(Path("a/b/c"), Path("a/b/c/d/e"))
+    Path('d/e')
+    >>> relative_to(Path("a/b/c/d/e"), Path("a/b/c/d/f/g"))
+    Path('../../f/g')
+    """
+    return Path(os.path.relpath(path1, path2))
+
+
+if __name__ == "__main__":
+    for app_path in static_preview_app_files(rootdir / "components"):
+        html_path = infile_to_outfile(app_path)
+
+        # # Skip if up-to-date, unless -f flag is passed
+        # if html_path.stat().st_mtime > (rootdir / app_path).stat().st_mtime:
+        #     continue
+
+        print(f"{app_path}\n    Writing to {html_path}")
+        start_time = time.perf_counter()
+        render_static_preview(app_path, html_path, libdir)
+        end_time = time.perf_counter()
+        print(f"    Succeeded in {((end_time - start_time) * 1000):.2f}ms")
+
+    # As of this writing, there's a bug in the bootstrap dependency that ships with
+    # shiny in that it doesn't include all_files=True, so some font stuff is missing.
+    # This code works around the bug by copying the missing files from the shiny package
+    # to the static lib directory. When the bug is fixed, this code can be removed.
```

Comment:
Is this a bug report? 😆 If you haven't fixed this yet, I can handle the py-shiny change.

Reply:
https://github.com/posit-dev/py-shiny/pull/1342

Reply:
Thank you!